test-data/unit/check-enum.test

```diff
@@ -951,3 +951,119 @@ class A:
         self.b = Enum("x", [("foo", "bar")])  # E: Enum type as attribute is not supported
 
 reveal_type(A().b)  # N: Revealed type is 'Any'
+
+[case testEnumReachabilityWithChaining]
+from enum import Enum
+
+class Foo(Enum):
+    A = 1
+    B = 2
+
+x: Foo
+y: Foo
+
+if x is y is Foo.A:
+   reveal_type(x)   # N: Revealed type is 'Literal[__main__.Foo.A]'
+   reveal_type(y)   # N: Revealed type is 'Literal[__main__.Foo.A]'
+else:
+   reveal_type(x)   # N: Revealed type is 'Literal[__main__.Foo.B]'
+   reveal_type(y)   # N: Revealed type is 'Literal[__main__.Foo.B]'
+reveal_type(x)      # N: Revealed type is '__main__.Foo'
+reveal_type(y)      # N: Revealed type is '__main__.Foo'
+
+if x is Foo.A is y:
+   reveal_type(x)   # N: Revealed type is 'Literal[__main__.Foo.A]'
+   reveal_type(y)   # N: Revealed type is 'Literal[__main__.Foo.A]'
+else:
+   reveal_type(x)   # N: Revealed type is 'Literal[__main__.Foo.B]'
+   reveal_type(y)   # N: Revealed type is 'Literal[__main__.Foo.B]'
+reveal_type(x)      # N: Revealed type is '__main__.Foo'
+reveal_type(y)      # N: Revealed type is '__main__.Foo'
+
+if Foo.A is x is y:
+   reveal_type(x)   # N: Revealed type is 'Literal[__main__.Foo.A]'
+   reveal_type(y)   # N: Revealed type is 'Literal[__main__.Foo.A]'
+else:
+   reveal_type(x)   # N: Revealed type is 'Literal[__main__.Foo.B]'
+   reveal_type(y)   # N: Revealed type is 'Literal[__main__.Foo.B]'
+reveal_type(x)      # N: Revealed type is '__main__.Foo'
+reveal_type(y)      # N: Revealed type is '__main__.Foo'
+
+[builtins fixtures/primitives.pyi]
+
+[case testEnumReachabilityWithChainingDisjoint]
+# flags: --warn-unreachable
+from enum import Enum
+
+class Foo(Enum):
+    A = 1
+    B = 2
+
+    # Used to divide up a chained comparison into multiple identity groups
+    def __lt__(self, other: object) -> bool: return True
+
+x: Foo
+y: Foo
+
+# No conflict
+if x is Foo.A < y is Foo.B:
+    reveal_type(x)  # N: Revealed type is 'Literal[__main__.Foo.A]'
+    reveal_type(y)  # N: Revealed type is 'Literal[__main__.Foo.B]'
+else:
+    reveal_type(x)  # N: Revealed type is 'Literal[__main__.Foo.B]'
+    reveal_type(y)  # N: Revealed type is 'Literal[__main__.Foo.A]'
+reveal_type(x)      # N: Revealed type is '__main__.Foo'
+reveal_type(y)      # N: Revealed type is '__main__.Foo'
+
+# The standard output when we end up inferring two disjoint facts about the same expr
+if x is Foo.A and x is Foo.B:
+    reveal_type(x)  # E: Statement is unreachable
+else:
+    reveal_type(x)  # N: Revealed type is '__main__.Foo'
+reveal_type(x)      # N: Revealed type is '__main__.Foo'
+
+# TODO: This should behave in the same way as above.
+# However, unlike the above, we currently don't progressively update the type of 'x' as
+# we check each individual comparison. So, when we do 'x is Foo.B', mypy still thinks
+# 'x' is of type 'Foo', which is why we get the below faulty result.
```

Comment:
Would it be possible to fix this by merging the groups if they contain expressions with the same literal hash and literal level `LITERAL_TYPE`? I mean my guess is that this fails because there are two groups in `simplified_operator_list` that contain `x`.

Reply:
I tried implementing this suggestion, and it seems to work!

It did make the grouping algorithm more complicated though -- the line-count roughly doubled in size, I think. I also tried implementing a more direct/naive "merge the groups after making them" approach, and that ended up being similarly complex.

Not sure if that's something we're ok with or not: this feels like a lot of code for what I suspect is ultimately a very rare edge case. Maybe it might be better to remove the changes I made and keep this TODO to try and help minimize the maintenance burden of these changes? LMK what you think.